pkg/ipcache/ipcache.go

```diff
@@ -259,15 +265,20 @@ func (ipc *IPCache) updateNamedPorts() (namedPortsChanged bool) {
 			npm[name][port] = struct{}{}
 		}
 	}
-	namedPortsChanged = !npm.Equal(ipc.namedPorts)
+	namedPortsChanged = !npm.Equal(old)
 	if namedPortsChanged {
-		// swap the new map in
-		if len(npm) == 0 {
-			ipc.namedPorts = nil
-		} else {
-			ipc.namedPorts = npm
-		}
+		// atomically swap the new map in
+		ipc.namedPorts.Store(&npm)
 	}
+
+	// Set namedPortsChanged to false if they have not (yet) been requested.
+	// This avoids triggering policy updates if named ports changed, but no
+	// policy actually needs them. We still pre-compute the namedPorts map,
+	// so we don't have to acquire the IPCache lock in GetNamedPorts
+	if !ipc.needNamedPorts.Load() {
+		namedPortsChanged = false
+	}
```

Comment:
If we don't trigger policy updates due to the named port changes, then how does the policy evaluation occur in order to set the need for named ports?

Reply:
This _feels_ like it may introduce a behaviour where someone changes a named port on a Pod, and it doesn't take effect until later on when policy evaluation is triggered for some other unrelated reason.

Reply:
Yeah, I also had that thought. But note that the old code also has the behavior (this if statement here just used to be the first statement of this function, the only difference now is that we're always computing the namedPorts map now).

I'll investigate.

Reply:
> If we don't trigger policy updates due to the named port changes, then how does the policy evaluation occur in order to set the need for named ports?

So the way this seems to work is that `ipc.needNamedPorts` is false as long as there is no policy referencing named ports. So it's safe to not trigger policy updates, because at that point no policy was using them. Once a policy has named ports, the policy engine will call "GetNamedPorts", and thereby causing any subsequent `IPCache.{Upsert,Delete}` calls to return 'namedPortsChanged=true', enabling callers to trigger policy updates when needed.

I've left a source code comment indicating that as well. Note that this behavior was already present before this PR. This PR does not change any externally visible behavior of IPCache around named ports (besides the fact that internally it no longer acquires a mutex for `GetNamedPorts).